src/librustc_mir/interpret/place.rs

```diff
@@ -595,6 +595,14 @@ where
             }
 
             StaticKind::Static => {
+                //if the first frame on the stack isn't a static item, then we shouldn't
+                //eval any static places
+                if let ty::InstanceDef::Item(item_def_id) = self.stack[0].instance.def {
+                    if !self.tcx.is_static(item_def_id) {
+                        trace!("eval_static_to_mplace: can't eval static in constant");
+                        throw_unsup!(ReadOfStaticInConst);
+                    }
+                }
```

Comment:
This seems like the wrong time and place for such a check to me.

IMO it would be much more reasonable to do this check in memory's `get_static_alloc`. Also I think looking at the top frame is quite hacky; instead there should be a machine hook indicating whether access to statics is allowed which CTFE can then implement appropriately.

Also, if this error is never caught, `throw_unsup_format` is the better choice. I think having so many error variants in the Miri engine is mostly annoying.

Reply:
Ah, but maybe this does get caught because const_prop implicitly catches all errors.

Then having the variant is correct, but it should be added above the "-- Everything below is not classified yet --" separator. We shouldn't add *new* unclassified error variants but instead just classify the new ones correctly. ;)

Reply:
Also Cc https://github.com/rust-rfcs/const-eval/pull/27 where I am speculating about adding exactly this kind of check to the engine.

Note, however, that *immutable* statics *can* be accesses from a constant just fine. We currently just reject all statics because that's simpler. So if there really is a problem with evaluating statics in general, even immutable ones, I think that's a bug and should be fixed.

Reply:
Is this the kind of classification you're looking for? 

https://github.com/rust-lang/rust/blob/c8eefb66b789437ee156c990ef746d3f324e48d1/src/librustc/mir/interpret/error.rs#L372-L378

Reply:
Yes, that looks about right (but see the other discussion for which category `UninitializedLocal` should be in).

Also see https://github.com/rust-lang/rust/pull/64541.